File: 1.two-sum.java
// import java.util.Map;

/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start
public class Solution {
    public int[] twoSum(int[] nums, int target) {
		int[] ans = new int[2];
		HashMap<Integer, Integer> map = new HashMap<>();

		for (int index = 0; index < nums.length; index++) {
			if (!map.containsKey(nums[index])) {
				map.put(target - nums[index], index);
			} else {
                ans[0] = map.get(nums[index]);
				ans[1] = index;
				break;
                
			}
		}
        
		return ans;
	}
}
// @lc code=end

File: cpp/1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

// @lc code=start
// Note: First leetcode problem start
// use clang formatter
// include vector and std to avoid issue in the code, not need in solution
#include <vector>
using namespace std;

class Solution {
   public:
    // what is vector? int is a generic type in vector
    // can use double as well
    // vector is a dynamic array in C++ STL
    vector<int> twoSum(vector<int> &nums, int target) {
        // why pass nums by reference?
        // nums is passed by reference to avoid copying the entire vector
        int n = nums.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (nums[i] + nums[j] == target) {
                    // how to return vector of int?
                    // return {i, j};
                    // both working
                    std::vector<int> vec = {i, j};
                    return vec;
                }
            }
        }
        return {};
    }
};
// @lc code=end
